File: Templates/C#/IntegrationTesting/Project/UserService.cs
using System.Text.Json;

namespace IntegrationTesting;

public class UserService(IJsonPlaceholderClient jsonPlaceholderClient) : IUserService
{
    private readonly JsonSerializerOptions options = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public async Task<List<BaseUser>> GetAllAsync()
    {
        var content = await jsonPlaceholderClient.GetAsync("/users", null);
        return JsonSerializer.Deserialize<List<BaseUser>>(content, options) ?? [];
    }

    public async Task<BaseUser?> GetByIdAsync(string id)
    {
        var content = await jsonPlaceholderClient.GetAsync("/users", null);
        var allUsers = JsonSerializer.Deserialize<List<BaseUser>>(content, options) ?? [];
        var user = allUsers.FirstOrDefault(u => u.Id.ToString() == id);
        return user;
    }

    public async Task<List<AddressUser>> GetNearbyAsync(double lat, double lng, double miles)
    {
        var content = await jsonPlaceholderClient.GetAsync("/users", null);
        var allUsers = JsonSerializer.Deserialize<List<AddressUser>>(content, options) ?? [];
        return [.. allUsers.Where(user =>
            user.Address?.Geo != null &&
            DistanceUtils.CalculateDistance(lat, lng, user.Address.Geo.Lat, user.Address.Geo.Lng) <= miles
        )];
    }
}

File: Templates/JavaScript/src/user.service.js
import {calculateDistance} from './distanceUtils.js';

/** @typedef {import('./user.model.js').User User} */
/** @typedef {import('./user.model.js').AddressUser AddressUser} */

export class UserService {
    /**
     * @param {JsonPlaceholderClient} jsonPlaceholderClient
     */
    constructor(jsonPlaceholderClient) {
        this.client = jsonPlaceholderClient;
    }

    /**
     * @returns {Promise<Array<User>>}
     */
    async getAllAsync() {
        const content = await this.client.getAsync('/users');
        if (!content) {
            return [];
        }

        return /** @type Array<User> */ (JSON.parse(content));
    }

    /**
     * @param {string|number} id
     * @returns {Promise<User|null>}
     */
    async getByIdAsync(id) {
        const content =  await this.client.getAsync('/users');
        if (!content) {
            return null;
        }

        const allUsers = /** @type Array<User> */ (JSON.parse(content));
        return allUsers.find(user => user.id == id) || null;
    }

    /**
     * @param {number} lat
     * @param {number} lng
     * @param {number} miles
     * @returns {Promise<Array<AddressUser>>} Users within the spatial radius
     */
    async getNearbyAsync(lat, lng, miles) {
        const allUsers = /** @type Array<AddressUser> */(await this.getAllAsync());

        return allUsers.filter(user => {
            const geo = user.address?.geo;
            if (!geo || geo.lat === undefined || geo.lng === undefined) {
                return false;
            }

            const distance = calculateDistance(lat, lng, Number(geo.lat).toFixed(3), Number(geo.lng).toFixed(3));
            return distance <= miles;
        });
    }
}
